mathesar_ui/src/stores/databases.ts

```diff
@@ -54,3 +67,105 @@ export async function reloadDatabases(): Promise<PaginatedResponse<Database>> {
     return null;
   }
 }
+
+export type DbType = string;
+
+export interface MathesarType {
+  name: string,
+  identifier: string,
+  db_types: DbType[]
+}
+
+/**
+ * Considers a DbType to belong to the first MathesarType that has it in its `db_types` set.
+ */
+export function determineMathesarType(
+  // eslint-disable-next-line @typescript-eslint/no-shadow
+  mathesarTypes: MathesarType[],
+  dbType: DbType,
+): MathesarType {
+  const mathesarTypeHasItAsTarget = (mt: MathesarType) => mt.db_types.includes(dbType);
+  return mathesarTypes.find(mathesarTypeHasItAsTarget);
+}
+
+export function choosePreferredDbTypeTarget(
+  // eslint-disable-next-line @typescript-eslint/no-shadow
+  mathesarType: MathesarType,
+): DbType {
+  switch (mathesarType.identifier) {
+    case 'number':
+      return 'NUMERIC';
+    case 'text':
+      return 'VARCHAR';
+    default:
+      throw new Error(`Database type target undefined for Mathesar type ${mathesarType.name}`);
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-shadow
+export function getMathesarTypeIcon(mathesarType: MathesarType): string {
+  switch (mathesarType.identifier) {
+    case 'number':
+      return '#';
+    case 'text':
+      return 'T';
+    default:
+      return '?';
+  }
+}
+
+export type DatabasesToMathesarTypes = Map<Database['id'], MathesarType[]>;
+
+/**
+ * Given a set of Database objects, queries the API for Mathesar types allowed on those databases
+ * in parallel, and constructs a mapping between database ids and sets of those Mathesar types.
+ */
+async function getDatabasesToMathesarTypes(
+  knownDatabases: Database[],
+):Promise<DatabasesToMathesarTypes> {
+  function getMathesarTypesForDatabase(db: Database) {
+    return getAPI<MathesarType[]>(`/databases/${db.id}/types`);
+  }
+
+  const promisesOfPairs = knownDatabases.map(
+    async (db) => pair(db.id, await getMathesarTypesForDatabase(db)),
+  );
+
+  const toMap = <A, B>(pairs: [A, B][]) => new Map<A, B>(pairs);
+
+  return Promise.all(promisesOfPairs).then(toMap);
+}
+
+/**
+ * A Readable containing the output of calling getDatabasesToMathesarTypes on the database set
+ * stored in the `databases` store.
+ */
+// eslint-disable-next-line operator-linebreak
+export const databasesToMathesarTypesStore: Readable<DatabasesToMathesarTypes> =
+  derived<Readable<DatabaseStoreData>, DatabasesToMathesarTypes>(
+    databases,
+    ($databaseStoreData, set) => {
+      const knownDatabases = $databaseStoreData.data;
+      if (knownDatabases && notEmpty(knownDatabases)) {
+        void getDatabasesToMathesarTypes(knownDatabases).then(set);
+      }
+    },
+    undefined,
+  );
```

Comment:
This is intriguing. I'll do some testing to verify what's going on. Good catch. My first intuition is that one of the dependency stores for the derived store is getting updated. It's really weird that a not-first subscribe could trigger the updater function on its own.

Reply:
This behaviour seems to occur only on derived stores, as far as I remember. It might also be a bug caused by a race condition in Svelte store implementation. Either way, this needs more analysis.